smp/src/smp.c

```diff
@@ -322,58 +316,60 @@ smp_process_request_packet(struct smp_streamer *streamer, void *req)
 
     rsp = NULL;
     valid_hdr = true;
+    rc = 0;
 
-    while (1) {
+    while (rc == 0) {
         rc = mgmt_streamer_init_reader(&streamer->mgmt_stmr, req);
         if (rc != 0) {
             valid_hdr = false;
-            break;
+            goto error;
         }
 
         /* Read the management header and strip it from the request. */
         rc = smp_read_hdr(streamer, &req_hdr);
         if (rc != 0) {
             valid_hdr = false;
-            break;
+            goto error;
         }
         mgmt_ntoh_hdr(&req_hdr);
         mgmt_streamer_trim_front(&streamer->mgmt_stmr, req, MGMT_HDR_SIZE);
 
         rsp = mgmt_streamer_alloc_rsp(&streamer->mgmt_stmr, req);
         if (rsp == NULL) {
             rc = MGMT_ERR_ENOMEM;
-            break;
+            goto error;
         }
 
         rc = mgmt_streamer_init_writer(&streamer->mgmt_stmr, rsp);
         if (rc != 0) {
-            break;
+            goto error;
         }
 
         /* Process the request payload and build the response. */
         rc = smp_handle_single_req(streamer, &req_hdr);
         if (rc != 0) {
-            break;
+            goto error;
         }
 
         /* Send the response. */
         rc = streamer->tx_rsp_cb(streamer, rsp, streamer->mgmt_stmr.cb_arg);
-        rsp = NULL;
         if (rc != 0) {
-            break;
+            goto error;
         }
 
         /* Trim processed request to free up space for subsequent responses. */
         mgmt_streamer_trim_front(&streamer->mgmt_stmr, req,
                                  smp_align4(req_hdr.nh_len));
-    }
 
-    if (rc != 0 && valid_hdr) {
-        smp_on_err(streamer, &req_hdr, req, rsp, rc);
-        return rc;
+error:
+        if (rc != 0 && valid_hdr) {
+            smp_on_err(streamer, &req_hdr, req, rsp, rc);
+        }
+
+        mgmt_streamer_free_buf(&streamer->mgmt_stmr, rsp);
```

Comment:
Won't this call to `mgmt_streamer_free_buf()` cause a double free?  The expectation is that `streamer->tx_rsp_cb()` consumes the response buffer (e.g., for Zephyr, this callback is `zephyr_smp_tx_rsp()`, which calls  `smp_bt_tx_pkt()`, which frees the buffer after use).

Reply:
In fact the current implementation causes double free.
If the `streamer->tx_rsp_cb()` consumes the response buffer, then why there is double call to mgmt_streamer_free_buf to free rsp and req buffers in smp_process_request_packet && smp_on_err? ;) In smp_on_err even NULL is assigned to req and rsp still calling mgmt_streamer_free_buf with this NULL parameter two times.

I know this patch is not perfect because it changes the concept by `streamer->tx_rsp_cb()` not consuming the buffer, so if you want to fix this in another way, please do or suggest me.

I thought that it might be cleaner if function that allocates the buffer frees it, I think that it's good to have this in place since this is not placed on any FIFO or sth, but freed in the same function context.

Reply:
I don't think there is a double free in the original implementation, because `rsp` gets set to null after the call to `tx_rsp_cb()`.  But I might be misreading what you mean.

I am still a bit confused about who is supposed to consume the response buffer in this patch.  Could you please clarify this a bit, specifically regarding the below listing:

```
        /* [1] Send the response. */
        rc = streamer->tx_rsp_cb(streamer, rsp, streamer->mgmt_stmr.cb_arg);
        if (rc != 0) {
            goto error;
        }

        /* Trim processed request to free up space for subsequent responses. */
        mgmt_streamer_trim_front(&streamer->mgmt_stmr, req,
                                 smp_align4(req_hdr.nh_len));

error:
        if (rc != 0 && valid_hdr) {
            smp_on_err(streamer, &req_hdr, req, rsp, rc);
        }

        /* [2] */
        mgmt_streamer_free_buf(&streamer->mgmt_stmr, rsp);
```

(I have added `[1]` and `[2]` labels in comments)

Does the response buffer get freed in `[1]`?  If so, doesn't it get freed a second time in `[2]`?

Reply:
It is not freed in [1]. [2] frees it. My intention is to change tx_rsp_cb to *not* consume the buffer.

Reply:
Ah, I see.  I think that change creates a problem, at least for Mynewt.  In Mynewt, the transmit callback doesn't actually transmit the buffer, it just adds it to a queue.  The queue is processed later (either by the same task or another), the buffer is transmitted and freed.  The transmit callback is a "hand off", so the mcumgr code cannot assume the response buffer is safe to free after the callback returns.

Maybe I need to understand the root problem a little better.  Sorry if this has already been explained, but would you mind clarifying the problem that this solves?